engine/variableValue.h

```diff
@@ -139,19 +141,28 @@ namespace minsky
       setXVector(std::move(xv));
       return d;
     }
-    size_t numElements() const {
+    // For feature 47
+    std::vector<size_t> index;
+    
+    double sparsityRatio=0.0;       
+    
+    size_t numDenseElements() const {                                                
       size_t s=1;
       for (auto& i: xVector) s*=i.size();
       return s;
     }
-
-    template <class T>
-    void setXVector_(T x) {
-      size_t prevNumElems=numElements();
-      m_xVector=x;
-      if (idx()==-1 || prevNumElems<numElements())
-        allocValue();
-    }
+    
+    size_t numSparseElements = 0.0;
```

Comment:
As mentioned, it would be better to carry this information as index.size(). You can add an accessor method if desired for stylistic reasons, though.

Reply:
Done, there is now a new member in variableValue called dataSize():

    size_t dataSize() const {
		return index.size() ? index.size(): numDenseElements();
	}